coffee-backend/src/main/kotlin/chat/letscoffee/config/SecurityConfig.kt

```diff
@@ -0,0 +1,123 @@
+import org.springframework.security.authentication.AuthenticationManager
+import org.springframework.security.config.BeanIds
+import org.springframework.security.config.annotation.authentication.builders.AuthenticationManagerBuilder
+import org.springframework.security.config.annotation.method.configuration.EnableGlobalMethodSecurity
+import org.springframework.security.config.annotation.web.builders.HttpSecurity
+import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
+import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter
+import org.springframework.security.config.http.SessionCreationPolicy
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
+import org.springframework.security.crypto.password.PasswordEncoder
+import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter
+
+@Configuration
+@EnableWebSecurity
+@EnableGlobalMethodSecurity(securedEnabled = true, jsr250Enabled = true, prePostEnabled = true)
+class SecurityConfig : WebSecurityConfigurerAdapter() {
+    @Autowired
+    private val customUserDetailsService: CustomUserDetailsService? = null
+    @Autowired
+    private val customOAuth2UserService: CustomOAuth2UserService? = null
+    @Autowired
+    private val oAuth2AuthenticationSuccessHandler: OAuth2AuthenticationSuccessHandler? = null
+    @Autowired
+    private val oAuth2AuthenticationFailureHandler: OAuth2AuthenticationFailureHandler? = null
+    @Autowired
+    private val httpCookieOAuth2AuthorizationRequestRepository: HttpCookieOAuth2AuthorizationRequestRepository? = null
+
+    @Bean
+    fun tokenAuthenticationFilter(): TokenAuthenticationFilter {
+        return TokenAuthenticationFilter()
+    }
+
+    /*
+      By default, Spring OAuth2 uses HttpSessionOAuth2AuthorizationRequestRepository to save
+      the authorization request. But, since our service is stateless, we can't save it in
+      the session. We'll save the request in a Base64 encoded cookie instead.
+    */
+    @Bean
+    fun cookieAuthorizationRequestRepository(): HttpCookieOAuth2AuthorizationRequestRepository {
+        return HttpCookieOAuth2AuthorizationRequestRepository()
+    }
+
+    @Throws(Exception::class)
+    public override fun configure(authenticationManagerBuilder: AuthenticationManagerBuilder) {
+        authenticationManagerBuilder
+                .userDetailsService(customUserDetailsService)
+                .passwordEncoder(passwordEncoder())
+    }
+
+    @Bean
+    fun passwordEncoder(): PasswordEncoder {
+        return BCryptPasswordEncoder()
+    }
+
+    @Bean(BeanIds.AUTHENTICATION_MANAGER)
+    @Throws(Exception::class)
+    override fun authenticationManagerBean(): AuthenticationManager {
+        return super.authenticationManagerBean()
+    }
+
+    @Throws(Exception::class)
+    override fun configure(http: HttpSecurity) {
+        http
```

Comment:
Don't really know what's happing in this builder, but it seems quite error-prone to have to define how all possible endpoints in the application should behave in a single file. The behaviour of an endpoint should be defined where the endpoint is defined. Is it possible to do this another way? Or not have it at all?

Reply:
This is just for security endpoints, as well as add basic Token filtering before each request.